### PR TITLE
distsql: copy prefix key in index skip table reader 

### PR DESCRIPTION
**Pull request by asubiotto**
This is necessary because the table reader goes on to modify the key
returned from `PartialKey` in place, which would corrupt the current row
to return since the key is unsafe.

This commit also changes test descriptions to the more conventional
TestName/AndSubtestName in the index skip table reader tests and
shortens them to be able to run specific subtests more easily (less
typing).

Release note: None

### CODE REVIEW

**Comment by asubiotto** [comment]
Also deleted some duplicate tests.

**Review by rohany — CHANGES_REQUESTED** [review]
Thanks for cleaning up the tests alfonso!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-:-LmBWyBKD4YqADS1ai7I:b-kwzd9c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @ridwanmsharif, and @rohany)

___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmBV-cn9nb-Ey445Chv:-LmBV-coBwBmUNLecfOE:b-ogn93o) ([raw file](https://github.com/cockroachdb/cockroach/blob/678af1568748efb83390a2af1960caf6760a3f30/pkg/sql/distsqlrun/index_skip_table_reader.go#L204)):*
> ```Go
> 			// If this is a forward scan, skip to the end of the current prefix's
> 			// keyspace on the next seek.
> 			t.spans[t.currentSpan].Key = prefixKey.PrefixEnd()
> ```

If we want to not modify the current key we should just make a copy here, rather than using prefixEnd. I'm not sure PrefixEnd guards against null values being in every other position in the index.

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ridwanmsharif and @rohany)

___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmBV-cn9nb-Ey445Chv:-LmBYIOS9CfsPVqbDzmr:b89j0qy) ([raw file](https://github.com/cockroachdb/cockroach/blob/678af1568748efb83390a2af1960caf6760a3f30/pkg/sql/distsqlrun/index_skip_table_reader.go#L204)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

If we want to not modify the current key we should just make a copy here, rather than using prefixEnd. I'm not sure PrefixEnd guards against null values being in every other position in the index.
</blockquote></details>

Good point, done. We should probably have a test that checks this case. We should also maybe just increment `currentSpan` if the prefix key is maximal?

**Review by ridwanmsharif — COMMENTED** [review]
The changes LGTM! Thanks for fixing this! I'll defer to @rohany for the ✅ cause I'm not too familiar with this area

**Review by rohany — APPROVED** [review]
LGTM with a nit + question

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-:-LmCKP1l6-JUa6Vwr2iH:b-gj4ngs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmBV-cn9nb-Ey445Chv:-LmCJgVg9q9f7o9YPTxk:b3p8egy) ([raw file](https://github.com/cockroachdb/cockroach/blob/678af1568748efb83390a2af1960caf6760a3f30/pkg/sql/distsqlrun/index_skip_table_reader.go#L204)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Good point, done. We should probably have a test that checks this case. We should also maybe just increment `currentSpan` if the prefix key is maximal?
</blockquote></details>

I thought the interleaved tests should have tested this, or some of the null tests, but I guess not. I think a table with an index of 3 columns, where the last 2 columns only contain nulls should test this behavior. Can you open an issue about that?

Do you mean if the key is all 0xff? We could check that here, but it will get caught in the next case anyway, where we check that the span is valid.
___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmCJbgb5yYHuWxCGHlE:-LmCJbgc5Or805JlbnjI:b8ur0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d7e4a9716f9ba596a7925c781cf0f6448128ec3/pkg/sql/distsqlrun/index_skip_table_reader.go#L203)):*
> ```Go
> 		if !t.reverse {
> 			// Modifying key in place would corrupt our current row, so make a copy.
> 			key = append([]byte(nil), key...)
> ```

[nit] would it be clearer to just use `copy`

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmCKbUI0zT3En-NwRdr:-LmCKbUI0zT3En-NwRds:b38fifr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d7e4a9716f9ba596a7925c781cf0f6448128ec3/pkg/sql/distsqlrun/index_skip_table_reader.go#L184)):*
> ```Go
> 		}
> 
> 		key, err := t.fetcher.PartialKey(t.keyPrefixLen)
> ```

Should you move the copy up here? In the case where we are doing a reverse scan we are using the key from the fetcher which could be a problem.

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmBV-cn9nb-Ey445Chv:-LmFRKMT1dYDEI5ZpfRL:b-s42iqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/678af1568748efb83390a2af1960caf6760a3f30/pkg/sql/distsqlrun/index_skip_table_reader.go#L204)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I thought the interleaved tests should have tested this, or some of the null tests, but I guess not. I think a table with an index of 3 columns, where the last 2 columns only contain nulls should test this behavior. Can you open an issue about that?

Do you mean if the key is all 0xff? We could check that here, but it will get caught in the next case anyway, where we check that the span is valid.
</blockquote></details>

Yes (maximal key/all 0xff). Aren't there cases where a max key is still valid (e.g. if the `EndKey` is a longer max key, I'm thinking in terms of a variable prefix length which is what ridwan is working on)? That's what I was worried about, but I guess you're saying there's something else? Why would the rest of the index (after the prefix length) matter? I'll merge it without modifying this logic but we should revisit this. I added a TODO.
___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmCKbUI0zT3En-NwRdr:-LmFQTTY6Rwv4hE6XtRM:b7fhxka) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d7e4a9716f9ba596a7925c781cf0f6448128ec3/pkg/sql/distsqlrun/index_skip_table_reader.go#L184)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Should you move the copy up here? In the case where we are doing a reverse scan we are using the key from the fetcher which could be a problem.
</blockquote></details>

The issue is that we're doing in-place modification of an unsafe key in the forward case, so it's only that that would corrupt the row. But I moved the copy up here and added a separate scope to disallow reuse of the `key` variable.
___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmCJbgb5yYHuWxCGHlE:-LmFQrKV7pKnic6jbi8h:bhy6p8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d7e4a9716f9ba596a7925c781cf0f6448128ec3/pkg/sql/distsqlrun/index_skip_table_reader.go#L203)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit] would it be clearer to just use `copy`
</blockquote></details>

True, in fact it's also more performant. I added a scratch variable though to avoid allocating on every lookup. I kept it as an append just to be sure we have enough capacity to copy the whole key (since it could have some extra bytes that we don't account for i.e. table/index ID).

**Review by rohany — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmBV-cn9nb-Ey445Chv:-LmFcBz518GPuvuWpOuS:b56k0z4) ([raw file](https://github.com/cockroachdb/cockroach/blob/678af1568748efb83390a2af1960caf6760a3f30/pkg/sql/distsqlrun/index_skip_table_reader.go#L204)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Yes (maximal key/all 0xff). Aren't there cases where a max key is still valid (e.g. if the `EndKey` is a longer max key, I'm thinking in terms of a variable prefix length which is what ridwan is working on)? That's what I was worried about, but I guess you're saying there's something else? Why would the rest of the index (after the prefix length) matter? I'll merge it without modifying this logic but we should revisit this. I added a TODO.
</blockquote></details>

I think we guarantee here that maximal keys are invalid here (by doing t.indexLen - t.keyPrefixLen + 1), but I'm not sure. We can talk about this in person when I come back and take a look at the issues -- it is tricky.
___
*[pkg/sql/distsqlrun/index_skip_table_reader.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39636#-LmCKbUI0zT3En-NwRdr:-LmFbFvcCkl8NuR13nDJ:b-or4wzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d7e4a9716f9ba596a7925c781cf0f6448128ec3/pkg/sql/distsqlrun/index_skip_table_reader.go#L184)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The issue is that we're doing in-place modification of an unsafe key in the forward case, so it's only that that would corrupt the row. But I moved the copy up here and added a separate scope to disallow reuse of the `key` variable.
</blockquote></details>

that makes sense -- great.

**Comment by asubiotto** [comment]
Looks like a storage flake: #39651

bors r=rohany

**Pull request merged**
